City-of-Helsinki/helsinki-paatokset
# PP-499: Add new method for fetching decisions from ES index.

This MR adds a new method for fetching office holder decisions form the Elasticsearch index instead of querying the database. This also adds some new fields to the index that are required for the decisions list.

This doesn't replace the old DB method yet with the new one. This is because we might want to benchmark and test this a bit more on the test environments and also because it requires reindexing all decisions with the new fields, which should be completed on production before this feature is enabled with !295

**To test**
- Checkout branch
- Run `make drush-cim drush-cr shell`
- Fetch data with the following commands (this will take a while)
```
drush ap:agg decisions --dataset=all --queue --decisionmaker_id=u511053001020vh1 -v
drush ap:agg decisions --dataset=all --queue --decisionmaker_id=u4804005010vh1 -v
drush queue:run ahjo_api_aggregation_queue -v
drush ap:ud -v
drush queue:run ahjo_api_retry_queue -v
drush ap:ud --logic=case -v
``` 
- Check that the decisions from 2017-2023 appear on these pages:
  - https://helsinki-paatokset.docker.so/fi/paattajat/aluekirjastopalvelujen-johtaja-aluekirjastopalvelut/paatokset
  - https://helsinki-paatokset.docker.so/fi/paattajat/yksikon-paallikko-asuntopalvelut/paatokset 
- Reindex the content:
``` 
drush sapi-rt decisions
drush sapi-i decisions
```
- Edit the following files by changing the function call from `getAgendasList` to `getAgendasListFromElasticSearch`
  - `public/modules/custom/paatokset_policymakers/paatokset_policymakers.module` line 171
  - `public/modules/custom/paatokset_submenus/src/Plugin/Block/AgendasSubmenuBlock.php` line 38
- Clear the cache with `make drush-cr`
- Reload the list pages and also open these URLs:
 - https://helsinki-paatokset.docker.so/fi/paattajat/aluekirjastopalvelujen-johtaja-aluekirjastopalvelut
 - https://helsinki-paatokset.docker.so/fi/paattajat/yksikon-paallikko-asuntopalvelut
 - The lists should work and the decisions should be in the correct order: sorted by year, descing order for the dates and numbers
- Compare the list to the version in production:
  - https://paatokset.hel.fi/fi/paattajat/yksikon-paallikko-asuntopalvelut/paatokset
  - https://paatokset.hel.fi/fi/paattajat/aluekirjastopalvelujen-johtaja-aluekirjastopalvelut/paatokset
  - The lists should have the same amount of decisions in the same order. The links might be a bit different, but they should take you to the same decisions
- Read code changes and make sure everything makes sense
- Compare the `getAgendasList` and `getAgendasListFromElasticSearch` methods
  - Is anything relevant missing from the new one? (The sorting in the ES version is much simpler because the section numbers are not treated as string)

**Benchmark performance**
- Run `drush php:cli` and then: 
```
$policymakerService = \Drupal::service('paatokset_policymakers');
$policymakerService->setPolicyMaker('u511053001020vh1');
```
- Test the original function: (paste the whole code section as one block so the start time and end time can be calculated properly)
```
$start_time = microtime(true);
$data = $policymakerService->getAgendasList(0, FALSE);
$end_time = microtime(true);
print count($data);
print $end_time - $start_time;
```
- Test the new function:
```
$start_time = microtime(true);
$data = $policymakerService->getAgendasListFromElasticSearch(0, FALSE);
$end_time = microtime(true);
print count($data);
print $end_time - $start_time;
```
- The amount of decisions found `count($data)` should always be the same. The elasticsearch version *should* be faster every time
- Test this on other policymakers: `$policymakerService->setPolicyMaker('u4804005010vh1');` 
- Find a few more more decisionmakers to test with from https://paatokset.hel.fi/fi/paattajat
  - Try to find some with a lot of decisions and some with very few decisions
  - You can get the decisionmakers ID by switching to the english languages version. It should be visible in the URL, for example: https://paatokset.hel.fi/en/decisionmakers/u51105300405050vh1
  - Exit out of the PHP command line interface back to the shell and add the decisions to the import queue with: `drush ap:agg decisions --dataset=all --queue --decisionmaker_id=ID -v` (replace the ID part)
  - Then run the queue with: `drush queue:run ahjo_api_aggregation_queue -v`
  - Then index the decisions with `drush sapi-i`
  - If there are more than a 1000 decisions, you can import them all by running multiple commands, for example:
```
drush ap:agg decisions --dataset=all --queue --decisionmaker_id=ID --start=2017-01-01T00:00:00 --end=2020-01-01T00:00:00 -v
drush ap:agg decisions --dataset=all --queue --decisionmaker_id=ID --start=2020-01-01T00:00:00 --end=2023-01-01T00:00:00 -v
```
- It's not necessary to get absolutely all of a specific office holder's decisions, but would be good to benchmark how this works with large amounts of data
- Run the same benchmarks a few times because the timing can vary
- Run the benchmarks while the aggregation queue and/or indexing function is running
- Also try changing the parameters of the command from `(0, FALSE)` to different limits `(100, FALSE)` and `(2, FALSE)` for example and to change the last parameter to TRUE (this groups the decisions by year, so the `count` will give different results).
- Are there any cases where the ES version is slower or the difference is minimal?
- Any cases where the different methods result in different amount of hits?
